sycl/source/detail/graph_impl.cpp

```diff
@@ -352,11 +352,76 @@ void graph_impl::removeRoot(const std::shared_ptr<node_impl> &Root) {
   MRoots.erase(Root);
 }
 
-std::shared_ptr<node_impl>
-graph_impl::add(const std::vector<std::shared_ptr<node_impl>> &Dep) {
-  // Copy deps so we can modify them
-  auto Deps = Dep;
+std::set<std::shared_ptr<node_impl>> graph_impl::getCGEdges(
+    const std::shared_ptr<sycl::detail::CG> &CommandGroup) const {
+  const auto &Requirements = CommandGroup->getRequirements();
+  if (!MAllowBuffers && Requirements.size()) {
+    throw sycl::exception(make_error_code(errc::invalid),
+                          "Cannot use buffers in a graph without passing the "
+                          "assume_buffer_outlives_graph property on "
+                          "Graph construction.");
+  }
+
+  if (CommandGroup->getType() == sycl::detail::CGType::Kernel) {
+    auto CGKernel =
+        static_cast<sycl::detail::CGExecKernel *>(CommandGroup.get());
+    if (CGKernel->hasStreams()) {
+      throw sycl::exception(
+          make_error_code(errc::invalid),
+          "Using sycl streams in a graph node is unsupported.");
+    }
+  }
+
+  // Add any nodes specified by event dependencies into the dependency list
+  std::set<std::shared_ptr<node_impl>> UniqueDeps;
+  for (auto &Dep : CommandGroup->getEvents()) {
+    if (auto NodeImpl = MEventsMap.find(Dep); NodeImpl == MEventsMap.end()) {
+      throw sycl::exception(sycl::make_error_code(errc::invalid),
+                            "Event dependency from handler::depends_on does "
+                            "not correspond to a node within the graph");
+    } else {
+      UniqueDeps.insert(NodeImpl->second);
+    }
+  }
+
+  // A unique set of dependencies obtained by checking requirements and events
+  for (auto &Req : Requirements) {
+    // Look through the graph for nodes which share this requirement
+    for (auto &Node : MNodeStorage) {
+      if (Node->hasRequirementDependency(Req)) {
+        bool ShouldAddDep = true;
+        // If any of this node's successors have this requirement then we skip
+        // adding the current node as a dependency.
+        for (auto &Succ : Node->MSuccessors) {
+          if (Succ.lock()->hasRequirementDependency(Req)) {
+            ShouldAddDep = false;
+            break;
+          }
+        }
+        if (ShouldAddDep) {
+          UniqueDeps.insert(Node);
+        }
+      }
+    }
+  }
+
+  return std::move(UniqueDeps);
```

Comment:
Extreme nit but I think the `std::move` is not necessary here, as the compiler knows already that it can be moved, but the explicit move here actually prevents that move from being elided, as I understand.